src/fastpath/fastpath.c

```diff
@@ -6,15 +6,175 @@
+
+            /* Equivalent to cancel_ipc */
+            endpoint_t *ep_ptr;
+            ep_ptr = EP_PTR(thread_state_get_blockingObject(dest->tcbState));
+            endpoint_ptr_set_epQueue_head_np(ep_ptr, TCB_REF(dest->tcbEPNext));
+            if (unlikely(dest->tcbEPNext)) {
+                dest->tcbEPNext->tcbEPPrev = NULL;
+            } else {
+                endpoint_ptr_mset_epQueue_tail_state(ep_ptr, 0, EPState_Idle);
+            }
+
+
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+            ksKernelEntry.is_fastpath = true;
+#endif
+            if (!dest->tcbSchedContext) {
+                schedContext_donate(sc, dest);
+            }
+            assert(dest->tcbSchedContext);
+
+            /* Enqueue signalled but continue executing signaller */
+            setRegister(dest, badgeRegister, badge);
+            thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+            tcbSchedAppend(dest);
+            restore_user_context();
+            UNREACHABLE();
+        } else {
+            ntfn_set_active(ntfnPtr, badge);
+            restore_user_context();
+            UNREACHABLE();
+        }
+        break;
+    }
+    case NtfnState_Waiting: {
+
+        /* Signal to higher prio thread is NOT fastpathed*/
+        if (NODE_STATE(ksCurThread)->tcbPriority < dest->tcbPriority) {
+            slowpath(SysSend);
+        }
+
+        /* Basically equivalent to maybeDonateSchedContext. Check whether the thread already has
+         * a SC or if one can be donated from the notification. If neither is true, go to
+         * slowpath */
+        if (!dest->tcbSchedContext) {
+            sc = SC_PTR(notification_ptr_get_ntfnSchedContext(ntfnPtr));
+            if (sc == NULL || sc->scTcb != NULL) {
+                slowpath(SysSend);
+            }
+        }
+
+        /*  Point of no return */
+
+        tcb_queue_t ntfn_queue;
+        ntfn_queue.head = (tcb_t *)notification_ptr_get_ntfnQueue_head(ntfnPtr);
+        ntfn_queue.end = (tcb_t *)notification_ptr_get_ntfnQueue_tail(ntfnPtr);
+
+        ntfn_queue = tcbEPDequeue(dest, ntfn_queue);
+
+        notification_ptr_set_ntfnQueue_head(ntfnPtr, (word_t)ntfn_queue.head);
+        notification_ptr_set_ntfnQueue_tail(ntfnPtr, (word_t)ntfn_queue.end);
+
+        if (!ntfn_queue.head) {
+            notification_ptr_set_state(ntfnPtr, NtfnState_Idle);
+        }
+
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+        ksKernelEntry.is_fastpath = true;
+#endif
+        if (!dest->tcbSchedContext) {
+            schedContext_donate(sc, dest);
+        }
+        assert(dest->tcbSchedContext);
+
+        /* Enqueue signalled but continue executing signaller */
+        setRegister(dest, badgeRegister, badge);
+        thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+        tcbSchedAppend(dest);
+        restore_user_context();
+        UNREACHABLE();
+    }
+    default:
+        slowpath(SysSend);
```

Comment:
I don't believe that this path can never be taken, and so should probably either be a panic() or have the `default:` case removed.

Reply:
explicit panic is probably advisable if it can't be taken

Reply:
That's true. The reason I left it like this is because I thought error handling should be left to the slowpath because we don't care about performance in this case, but since this would be more like corrupted kernel state than just an error, it  should never happen and a panic can be okay?

Reply:
I'll remove the default altogether.